src/background/Wallet/persistence.ts

```diff
@@ -42,25 +69,130 @@ export class WalletStore extends PersistentStore<WalletStoreState> {
+  async save(id: string, credentials: Credentials, record: WalletRecord) {
     if (this.lastRecord === record) {
       return;
     }
-    await this.encryptAndSave(id, encryptionKey, record);
+    await this.encryptAndSave(id, credentials, record);
+  }
+
+  async createBackup(id: string) {
+    /**
+     * Accessing user is a cross-concern, but this is the only way
+     * to make our backup truly atomic and independent:
+     * encrypted record relies on `salt` stored in the user object,
+     * and for a robust backup recovery it's best to store this object
+     * together with the encrypted record
+     */
+    const user = await Account.readCurrentUser();
+    const record = (await this.getSavedState())[id];
+    invariant(record, `Record not found for id: ${id}`);
+    invariant(user && user.id === id, `User not found for id: ${id}`);
+    return this.setState({
+      ...this.state,
+      [`backup:${id}`]: stringifyBackup({ user, record }),
+    });
+  }
+
+  async restoreFromBackup(id: string) {
+    const key = `backup:${id}`;
+    const state = await this.getSavedState();
+    const saved = state[key];
+    invariant(saved, `Backup not found for id: ${id}`);
+    const { user, record } = parseBackup(saved);
+    await Promise.all([
+      this.setState((state) =>
+        produce(state, (draft) => {
+          draft[id] = record;
+          delete draft[key];
+        })
+      ),
+      Account.writeCurrentUser(user),
+    ]);
+  }
+
+  async restoreFromAnyBackup() {
+    const state = await this.getSavedState();
+    const key = Object.keys(state).find((key) => key.startsWith('backup:'));
+    if (key) {
+      await this.restoreFromBackup(key.split(':')[1]);
+    } else {
+      throw new Error('No backups found');
+    }
+  }
+
+  async clearBackup(id: string) {
+    return this.setState((state) =>
+      produce(state, (draft) => {
+        const key = `backup:${id}`;
+        delete draft[key];
+      })
+    );
+  }
+
+  /**
+   * Executes an operation with a backup and an automatic recovery.
+   * Guarantees atomicity by restoring to the previous state if the operation fails.
+   */
+  async withBackup(id: string, operation: () => Promise<unknown>) {
+    await this.createBackup(id);
+    try {
+      await operation();
+      await this.clearBackup(id);
+    } catch (error) {
+      try {
+        await this.restoreFromBackup(id);
+        emitter.emit('globalError', {
+          name: 'internal_error',
+          message: 'Atomic wallet update failed. Restored from backup.',
+        });
+        console.log('Successfully restored wallet record from backup'); // eslint-disable-line no-console
+      } catch {
+        emitter.emit('globalError', {
+          name: 'internal_error',
+          message: 'Atomic wallet update failed. Restore from backup failed.',
+        });
+        throw new InternalBackupError(getError(error), { didRestore: false });
```

Comment:
Do we need this param for this Error? It looks like we only use it with didRestore: false